File: typescript/src/day09.ts
export default function day09(data: string) {
  const ns = data.split('\n').map(Number);

  const answer1 = part1(ns, 25);  
  const answer2 = part2(answer1, ns);

  console.log('===== Day 9 =====')
  console.log('The first wekness found is', answer1);
  console.log('The sum of min and max of the secret sequence is', answer2);
}

function part1(ns: number[], preamble: number): number {
  let i = preamble;
  while (true) {
    const range = ns.slice(i-preamble, i);
    const hasSum = findSum(ns[i], range);
    if (!hasSum) return ns[i];
    i++;
  }
}

function findSum(n: number, ns: number[]): boolean {
  for (let i = 0; i < ns.length; i++) {
    for(let j = 0; j < ns.length; j++) {
      if (i !== j && ns[i] + ns[j] === n) return true;
    }
  }
  return false;
}

function part2(target: number, ns: number[]): number {
  const range = findContiguousSum(target, ns);
  return Math.min(...range) + Math.max(...range);
}

function findContiguousSum(n: number, ns: number[]): number[] {
  for(let i = 0; i < ns.length; i++) {
    let sum = 0;
    let j = i;
    while (sum < n) {
      sum += ns[j];
      j++;
    }
    if (sum === n) return ns.slice(i, j);
  }
  return [];
}

File: haskell/src/Day09.hs
module Day09 (day09) where

import Data.Vector (Vector, (!), fromList)
import Data.Maybe (mapMaybe)

day09 :: IO ()
day09 = do
    s <- readFile "../data/day09.txt"
    let ns = parseNumbers s
    let answer1 = firstWeakness ns 25
    let answer2 = part2 ns answer1
    putStrLn "\n===== Day 9 ====="
    putStrLn $ "The first number exposing the weakness is " ++ show answer1
    putStrLn $ "The checksum for the sequence adding up to the weak number is " ++ show answer2

firstWeakness :: Vector Int -> Int -> Int
firstWeakness ns preambleSize =
    (!) ns $ head $ filter (isWeakness ns) [preambleSize .. (length ns - 1)]

isWeakness :: Vector Int -> Int -> Bool
isWeakness ns pos = 
    let
        indexPairs = [(i, j) | i <- [(pos-25) .. (pos-2)],
                               j <- [(i+1) .. (pos-1)] ]
        pairSums = map (\(i, j) -> (ns ! i) + (ns ! j)) indexPairs
    in
        (ns ! pos) `notElem` pairSums

part2 :: Vector Int -> Int -> Int
part2 ns target =
    let
        seq = findSequenceWithSum ns target
    in
        minimum seq + maximum seq

findSequenceWithSum :: Vector Int -> Int -> [Int]
findSequenceWithSum ns target =
    head $ mapMaybe (sequenceToTarget ns target) [0 .. (length ns - 1)]


sequenceToTarget :: Vector Int -> Int -> Int -> Maybe [Int]
sequenceToTarget ns target start =
    go start [] target
    where
        go _ result 0 = Just result
        go i result target = if target < 0 
                             then Nothing 
                             else go (i+1) ((ns ! i) : result) (target - (ns ! i))

parseNumbers :: String -> Vector Int
parseNumbers = fromList . map read . lines